features/lwipstack/lwip-sys/arch/lwip_sys_arch.c

```diff
@@ -174,11 +177,11 @@ void sys_mbox_post(sys_mbox_t *mbox, void *msg) {
 
     int state = osKernelLock();
 
-    mbox->queue[mbox->post_idx % MB_SIZE] = msg;
-    mbox->post_idx += 1;
+    mbox->queue[mbox->post_idx] = msg;
+    mbox->post_idx = (mbox->post_idx + 1) % MB_SIZE;
 
     osEventFlagsSet(mbox->id, SYS_MBOX_FETCH_EVENT);
-    if (mbox->post_idx - mbox->fetch_idx == MB_SIZE-1)
+    if (SUB_MOD(mbox->post_idx, mbox->fetch_idx, MB_SIZE) >= MB_SIZE-1)
```

Comment:
I think it would be a bit simpler to do this by rearranging to avoid the subtraction. How about:

    if ((mbox->post_idx + 1) % MB_SIZE == mbox->fetch_idx)

Seems clear enough - if the next increment would take us to the fetch position, we're full. (Because that increment would make them equal, and be indistinguishable from "empty").

Reply:
Good remark. This check can be expressed in 3 different ways all derived from the same identity (`p` states for post, `f` - fetch)

(1) `p - f = MB_SIZE - 1`, where `p`, `f` belongs to modulus group `G(MB_SIZE)`

So:

(2) `p = f + MB_SIZE - 1` and since `MB_SIZE - 1 = -1`

(3) `p  = f - 1` <=> `p + 1 = f`

I decided to use (1) since the original code used subtraction (aka elements distance) which seems intuitive for me. Your proposal is (3), but note that the check can also be expressed via (2) as:

```c
if (mbox->post_idx == (mbox->fetch_idx + MB_SIZE - 1) % MB_SIZE)
```

Which also avoids modulus subtraction. What is more clear - it's subjective. In terms on computational overhead: (1), (2), (3) so your choice wins.

I'm afraid I don't follow your semaphore note. `EventFlag` in this code acts as a classical semaphore approach, open if mailbox's indexes are contained in specific range.